#This R script is to run the R portion of the RNA automated kallisto pipeline 
#Start time used in conjunction with the last 3 lines of code to tell how long the code ran for 
#start_time = Sys.time()

library(tximport)
library(DESeq2)

#read in data from the python script associated with this automated pipeline
cmd_line <- commandArgs(TRUE)
out_dir <- toString(cmd_line[1])
#Taking the output path to the main folder and adding 2 variables to read and write out to the appropriate files
kal_out_dir <- paste(out_dir, "Kallisto_out", sep="/")
des_out_dir <- paste(out_dir, "DESeq2_out", sep="/")

#Preparing to run Tximport 
#Using the previous python script to create a name and condition table, this will read in the table to use for tximport
nam_cond_tab <- paste(out_dir, "name_condition_table.txt", sep="/")
sam <- read.table(nam_cond_tab, header=T, sep="\t")
#Save the file paths by the absolute path, using the folder names generated by the names in the Name column of the name_condition table
files <- file.path(kal_out_dir ,sam$Name, "abundance.h5")
#Reading in the transcript to gene conversion table and running the tximport command with it 
gene_tran_table <- toString(cmd_line[2])
gen_tra_tab <- read.table(gene_tran_table, header = F, sep="\t")
txi.kallisto <- tximport(files, type="kallisto", txOut=F, tx2gene = gen_tra_tab)

#If it is needed to run tximport under one environment and DESeq2 under another environment. Run tximport in 1 environment then save the data, move to second environment and load the data and resume the code 
#txisave <- paste(out_dir, "txi_kal_object.R", sep="/")
#save(txi.kallisto, file= txisave)
#load(txisave)

#Running DESeq2
#Rowname change is to prevent a potential error that says the row names are not of type characters 
rownames(txi.kallisto$counts) <- as.character(rownames(txi.kallisto$counts))
ddsTxi <- DESeqDataSetFromTximport(txi.kallisto, colData = sam, design = ~Condition)
rundata <- DESeq(ddsTxi)
res <- results(rundata)
#Write out the data in a tab delimited table file 
res_out <- paste(des_out_dir, "DESeq2_output.txt", sep="/")
write.table(res, file=res_out, quote=F, sep="\t")
#Creating a PCA plot of the data

#For the test cases the small amount of reads is giving an error here going to comment it out for now
vsd <- vst(rundata)
pcaout <- paste(des_out_dir, "DESeq2_PCA_plot.png", sep="/")
png(pcaout, height=800, width=800, res=150)
plotPCA(vsd, intgroup=c("Name","Condition"))
dev.off()

#This code will be used to calculate the ranking. It will calculate the -log10 of the pvalue and then will take the sign of the log2foldchange and will combine the two
logfun <- function(x){
-log10(x)
}
logfun3 <- function(x){
ifelse(is.na(x), x=0, -log10(x))
}
signfun <- function(y){
sign(y)
}
rankfun <- function(z,a){
z*a
}
#This code will run the previous functions and store the intermediates and rankings in a variable
tab <- data.frame("ID"= rownames(res), "pvalue" = res$pvalue, "log2foldchange"= res$log2FoldChange)
tab <- cbind(tab, "Log"= sapply(tab$pvalue, logfun))
tab <- cbind(tab, "Sign"= sapply(tab$log2foldchange, signfun))
tab <- cbind(tab, "Rank"= mapply(rankfun, tab$Log, tab$Sign))
#This transforms any Na values to 0 so that when it orders the rows by rank all the NA will appear in the middle between the positive and negative values 
tab$Rank[is.na(tab$Rank)] <- 0
fintab <- tab[order(tab$Rank, decreasing =T),]
tab_out <- paste(des_out_dir, "DESeq2_Ranked_table.txt", sep="/")
#Writing out a table of the rankings and intermediate values for each gene. Done in descending order so that the highest expression change are at the top and bottom of the file, top for highest increased expression and bottom for highest decreased expression
write.table(fintab, file = tab_out, quote =F, sep="\t", row.names=F) 
#This will create a variable called hold that will hold the top 25 up regulated and top 25 down regulated transcripts according to the ranking to print out to a file
hold <- head(fintab, 25)
hold <- rbind(hold, tail(fintab, 25))
sigout <- paste(des_out_dir, "Top_50_Ranked_Genes.txt", sep="/")
write.table(hold, file = sigout, quote=F, sep="\t", row.names=F)


#create plots for the following: histogram or violin plot of the log2 fold change,DESeq2 PCA plot, volcano plot of the -log10(pvalue)
#fintab #deseq2 ranked table
#This will create a volcano plot mapping the -log10(pvalue) vs the log2foldchange
out_nam <- paste(des_out_dir, "/DESeq2_pvallog2_volcano_plot.png", sep="/")
png(out_nam, height=800, width=800, res=150)
plot(fintab$log2foldchange, fintab$Log, xlab = "Log2FoldChange", ylab = "-Log10(pvalue)", main = "Volcano Plot for log2foldchange and Pvalue", pch= 20, cex= 0.5)
dev.off()
#This will create a histogram of the log2foldchange
out_namman <- paste(des_out_dir, "/DESeq2_log2foldchange_Histogram.png", sep="/")
png(out_namman, height=800, width=800, res=150)
hist(fintab$log2foldchange, breaks= 50, xlab = "Log2FoldChange", ylab = "Frequency", main = "Histogram for log2foldchange", col ="red")
dev.off()

#this plus the start_time at the top can be used to calculate the time it takes for this script to run
#end_time = Sys.time()
#tot_time = end_time - start_time
#print(tot_time)
